operators/pkg/instautoctrl/submission.go

```diff
@@ -0,0 +1,193 @@
+	Scheme             *runtime.Scheme
+	NamespaceWhitelist metav1.LabelSelector
+	ContainerEnvOpts   forge.ContainerEnvOpts
+	// This function, if configured, is deferred at the beginning of the Reconcile.
+	// Specifically, it is meant to be set to GinkgoRecover during the tests,
+	// in order to lead to a controlled failure in case the Reconcile panics.
+	ReconcileDeferHook func()
+}
+
+// SetupWithManager registers a new controller for InstanceSubmissionReconciler resources.
+func (r *InstanceSubmissionReconciler) SetupWithManager(mgr ctrl.Manager, concurrency int) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&clv1alpha2.Instance{}).
+		Owns(&batch.Job{}).
+		Named("instance-submission").
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: concurrency,
+		}).
+		Complete(r)
+}
+
+// Reconcile reconciles the status of the InstanceSnapshot resource.
+func (r *InstanceSubmissionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if r.ReconcileDeferHook != nil {
+		defer r.ReconcileDeferHook()
+	}
+
+	log := ctrl.LoggerFrom(ctx, "instance", req.NamespacedName)
+	dbgLog := log.V(utils.LogDebugLevel)
+	tracer := trace.New("reconcile", trace.Field{Key: "instance", Value: req.NamespacedName})
+	ctx = ctrl.LoggerInto(trace.ContextWithTrace(ctx, tracer), log)
+
+	if log.V(4).Enabled() {
+		defer tracer.Log()
+	} else {
+		defer tracer.LogIfLong(utils.LongThreshold())
+	}
+
+	// Get the instance object.
+	var instance clv1alpha2.Instance
+	if err := r.Get(ctx, req.NamespacedName, &instance); err != nil {
+		if !kerrors.IsNotFound(err) {
+			log.Error(err, "failed retrieving instance")
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	tracer.Step("instance retrieved")
+
+	// Skip if the instance has not to be submitted.
+	if proceed, err := r.CheckLabelSelectors(ctx, &instance); !proceed {
+		if err == nil {
+			dbgLog.Info("reconciliation skipped", "reason", "label selector not matching")
+		}
+		return ctrl.Result{}, err
+	}
+	tracer.Step("labels checked")
+
+	templateName := types.NamespacedName{
+		Namespace: instance.Spec.Template.Namespace,
+		Name:      instance.Spec.Template.Name,
+	}
+	var template clv1alpha2.Template
+	if err := r.Get(ctx, templateName, &template); err != nil {
+		log.Error(err, "failed retrieving the instance template", "template", templateName)
+		return ctrl.Result{}, err
+	}
+	tracer.Step("retrieved the instance template")
+
+	if len(template.Spec.EnvironmentList) != 1 {
+		return ctrl.Result{}, fmt.Errorf("only one environment per template is supported")
+	}
+
+	environment := template.Spec.EnvironmentList[0]
+	jobStatus, isInvalidConfig, err := r.EnforceInstanceSubmissionJob(ctx, &instance, &environment)
+
+	switch {
+	case isInvalidConfig: // the instance cannot be submitted due to invalid configuration
+		instance.SetLabels(forge.InstanceAutomationLabelsOnSubmission(instance.GetLabels(), false))
+		dbgLog.Info("instance submission aborted")
```

Comment:
Shouldn't this be an error?

Reply:
Not necessarily, if the destination url is not set because you don't want submission to occur it shouldn't be a proper error

Reply:
Ok, but then there shouldn't be the submission-requested label in the first place

Reply:
Thing is the termination controller is not aware of the thing. I should add the check over the environment also there in order to correct this behavior

Reply:
I feel it would be more intuitive to have that label set only in case termination is actually required, but if that is too complex also the current alternative seems fine to me